dev/set_matrix.py

```diff
@@ -0,0 +1,427 @@
+    False
+    """
+    # Ref.: https://www.python.org/dev/peps/pep-0440/#final-releases
+    return re.search(r"^\d+(\.\d+)+$", ver) is not None
+
+
+def select_latest_micro_versions(versions):
+    """
+    Selects the latest micro version in each minor version.
+
+    Examples
+    --------
+    >>> versions = {
+    ...     "1.3.0": "2020-01-01T00:00:00",
+    ...     "1.3.1": "2020-02-01T00:00:00",  # latest in 1.3
+    ...     "1.4.0": "2020-03-01T00:00:00",
+    ...     "1.4.1": "2020-04-01T00:00:00",
+    ...     "1.4.2": "2020-05-01T00:00:00",  # latest in 1.4
+    ... }
+    >>> select_latest_micro_versions(versions)
+    ['1.3.1', '1.4.2']
+    """
+    seen_minors = set()
+    res = []
+
+    for ver, _ in sorted(
+        versions.items(),
+        # sort by (minor_version, upload_time) in descending order
+        key=lambda x: (LooseVersion(x[0]).version[:2], x[1]),
+        reverse=True,
+    ):
+        minor_ver = tuple(LooseVersion(ver).version[:2])  # A set doesn't accept a list
+
+        if minor_ver not in seen_minors:
+            seen_minors.add(minor_ver)
+            res.insert(0, ver)
+
+    return res
+
+
+def filter_versions(versions, min_ver, max_ver, excludes=None):
+    """
+    Filter versions that satisfy the following conditions:
+
+    1. is newer than or equal to `min_ver`
+    2. shares the same major version as `max_ver` or `min_ver`
+    3. (Optional) is not in `excludes`
+
+    Examples
+    --------
+    >>> versions = {
+    ...     "0.1.0": "2020-01-01T00:00:01",
+    ...     "0.2.0": "2020-01-01T00:00:02",
+    ...     "1.0.0": "2020-01-01T00:00:00",
+    ...     "1.1.0": "2020-01-01T00:01:00",
+    ... }
+    >>> filter_versions(versions, "0.1.0", "0.2.0")  # fetch up to the latest in 0.x.y
+    {'0.1.0': ..., '0.2.0': ...}
+    >>> filter_versions(versions, "0.1.0", "1.0.0")  # fetch up to the latest in 1.x.y
+    {'0.1.0': ..., '0.2.0': ..., '1.0.0': ..., '1.1.0': ...}
+    >>> filter_versions(versions, "0.1.0", "1.0.0", excludes=["0.2.0"])
+    {'0.1.0': ..., '1.0.0': ..., '1.1.0': ...}
+    """
+    if excludes is None:
+        excludes = []
+
+    # prevent specifying non-existent versions
+    assert min_ver in versions
+    assert max_ver in versions
+    assert all(v in versions for v in excludes)
+
+    versions = {v: t for v, t in versions.items() if v not in excludes}
+    versions = {v: t for v, t in versions.items() if is_final_release(v)}
+
+    max_major = get_major_version(max_ver)
+    versions = {v: t for v, t in versions.items() if get_major_version(v) <= max_major}
+    versions = {v: t for v, t in versions.items() if LooseVersion(v) >= LooseVersion(min_ver)}
+
+    return versions
+
+
+def get_changed_flavors(changed_files, flavors):
+    """
+    Detects changed flavors from a list of changed files.
+
+    Examples
+    --------
+    >>> flavors = ["pytorch", "xgboost"]
+    >>> get_changed_flavors(["mlflow/pytorch/__init__.py", "mlflow/xgboost.py"], flavors)
```

Comment:
this is out of the scope of this PR but should we unify the file structure for flavors?

Current:

```
# both file and folder/__init__.py exist

xgboost.py
pytorch/__init__.py
keras.py
sklearn/__init__.py
```

Proposed:

```
# unify them to folder/__init__.py

xgboost/__init__.py
pytorch/__init__.py
keras/__init__.py
sklearn/__init__.py
```

Reply:
Sounds good too me as a follow-up change!